File: Src/DetailedSamples/CSharp/FTP-related/ClientFtp/LocalFolderTreeNode.cs
using System;
using System.Windows.Forms;
using Xceed.FileSystem;

namespace ClientFtp
{
	public class LocalFolderTreeNode : TreeNode
	{
		public LocalFolderTreeNode( AbstractFolder folder )
		{
      if( folder == null )
        throw new ArgumentNullException( "folder" );

      m_folder = folder;

      if( folder.IsRoot )
      {
        this.Text = folder.FullName;
      }
      else
      {
        this.Text = folder.Name;
      }

      this.ImageIndex = ( int ) FtpItemIconEnum.ClosedFolder;
      this.SelectedImageIndex = ( int ) FtpItemIconEnum.OpenedFolder;

      // Make sure a [+] appears, using a dummy node.
      this.Nodes.Add( new TreeNode( "Browsing folder contents..." ) );
    }

    public AbstractFolder Folder
    {
      get { return m_folder; }
    }

    public void UpdateContents()
    {
      // Make sure to stay in the same expanded state
      bool expanded = this.IsExpanded;
      AbstractFolder[] folders = null;

      try
      {
        folders = m_folder.GetFolders( false );
      }
      finally
      {
        // Make sure to empty even if we get an exception.
        this.Nodes.Clear();
      }

      foreach( AbstractFolder folder in folders )
      {
        this.Nodes.Add( new LocalFolderTreeNode( folder ) );
      }

      if( expanded )
      {
        this.Expand();
      }
    }

    public void FillList( ListView contents )
    {
      FileSystemItem[] items = null;

      try
      {
        items = m_folder.GetItems( false );
      }
      finally
      {
        // Make sure to empty even if we get an exception.
        contents.Items.Clear();
      }

      foreach( FileSystemItem item in items )
      {
        contents.Items.Add( new LocalListViewItem( item ) );
      }
    }

    private AbstractFolder m_folder = null;
	}
}

File: Src/DetailedSamples/Visual Basic .NET/FTP-related/ClientFtp/LocalFolderTreeNode.vb
Imports System.Windows.Forms
Imports Xceed.FileSystem

Public Class LocalFolderTreeNode
  Inherits TreeNode

  Public Sub New(ByVal folder As AbstractFolder)
    If folder Is Nothing Then
      Throw New ArgumentNullException("folder")
    End If

    m_folder = folder

    If folder.IsRoot Then
      Me.Text = folder.FullName
    Else
      Me.Text = folder.Name
    End If

    Me.ImageIndex = CInt(FtpItemIconEnum.ClosedFolder)
    Me.SelectedImageIndex = CInt(FtpItemIconEnum.OpenedFolder)

    ' Make sure a [+] appears, using a dummy node.
    Me.Nodes.Add(New TreeNode("Browsing folder contents..."))
  End Sub


  Public ReadOnly Property Folder() As AbstractFolder
    Get
      Return m_folder
    End Get
  End Property


  Public Sub UpdateContents()
    ' Make sure to stay in the same expanded state
    Dim expanded As Boolean = Me.IsExpanded
    Dim folders As AbstractFolder() = Nothing

    Try
      folders = m_folder.GetFolders(False)
    Finally
      ' Make sure to empty even if we get an exception.
      Me.Nodes.Clear()
    End Try

    Dim folder As AbstractFolder

    For Each folder In folders
      Me.Nodes.Add(New LocalFolderTreeNode(folder))
    Next folder

    If expanded Then
      Me.Expand()
    End If
  End Sub


  Public Sub FillList(ByVal contents As ListView)
    Dim items As FileSystemItem() = Nothing

    Try
      items = m_folder.GetItems(False)
    Finally
      ' Make sure to empty even if we get an exception.
      contents.Items.Clear()
    End Try

    Dim item As FileSystemItem
    For Each item In items
      contents.Items.Add(New LocalListViewItem(item))
    Next item
  End Sub


  Private m_folder As AbstractFolder = Nothing
End Class
